### PR TITLE
ghooks configuration

### PR DESCRIPTION
**Pull request by ta2edchimp**
This PR introduces the ability to configure ghooks more detailed (as originally proposed [here](https://github.com/gtramontina/ghooks/issues/28#issuecomment-169419267)).

In `package.json`:

``` JSON
  "config": {
    "ghooks": {
        "randomConfig": ["randomValue"],
        "hooks": {
          "pre-commit": "echo \"some pre-commit hook\"
        }
    }
```

Replaces the old way (still working, but considered _deprecated_):

``` JSON
  "config": {
    "ghooks": {
        "pre-commit": "echo \"some pre-commit hook\""
    },
    "ghooks-randomConfig": ["introduced for backwards compatibility with hooks configuration"]
  }
```

The idea behind this PR is, for example to prepare ghooks to be able to **optionally** exit the hooks with a non-zero code (an error) if `ghooks` could not be loaded successfully. Therefore, **I consider this PR as work in progress**, until that particular feature can be (safely, this time) enabled again.

So in order to re-implement an »exit with error« functionality, I'd like to know what way to approach this would be the most adequate:
1. upon `install`, check the config whether to »exit with error« and write the hooks' script accordingly
2. try to obtain the responsible `package.json` and read the config, **not** exiting with an error if nothing explicitly states so.

When going route 1, I'd suggest making the `install` script "re-runnable", by adding it as `bin.ghooks` to the `package.json`.

@gtramontina @kentcdodds I'd like to hear from you what your thoughts are about this proposal (if/when you got time for this — no need to hurry).

**Edit:** _If_ this gets merged eventually, I'd propose we call this a breaking change to trigger a major version bump?

### CODE REVIEW

**Comment by ta2edchimp** [comment]
Meh. CI builds seem to fail [due to a bug in `istanbul`](https://github.com/istanbuljs/istanbul-api/issues/2) 😞

File: test/config.test.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        'ghooks-exitWithError': expectedConfig.exitWithError,
+      },
+    })
+
+    const {getConfig} = proxyquire('../lib/config', stub)
+
+    expect(getConfig()).to.deep.equal(expectedConfig)
+  })
+
+  it('reads custom configuration and hooks (overriding the deprecated way)', () => {
```
Comment (kentcdodds) [inline test/config.test.js]: great coverage!

File: package.json [hunk 2]
```diff
@@ -67,6 +71,9 @@
   "babel": {
     "presets": [
       "es2015"
+    ],
+    "plugins": [
+      "transform-object-rest-spread"
```
Comment (kentcdodds) [inline package.json]: I would have just included `stage-2`, ¯\\\_(ツ)_/¯

File: lib/runner.js [hunk 3]
```diff
@@ -17,10 +16,8 @@ function hook(dirname, filename) {
 }
 
 function commandFor(hookName) {
-  const pkgFile = findup.sync(process.cwd(), 'package.json')
```
Comment (kentcdodds) [inline lib/runner.js]: nice to move this logic to somewhere isolated where it can be tested easier :+1:

**Comment by kentcdodds** [comment]
This looks awesome!

> If this gets merged eventually, I'd propose we call this a breaking change to trigger a major version bump?

There's technically nothing breaking about this right? Unless I misunderstand what's going on here, it looks like we're just adding the ability to configure things. Seems fine to me. When we remove the ability to do the old stuff then we can version bump. But it looks like we've got backward compat right now.

File: package.json (hunk 2 above)
Comment (ta2edchimp) [inline package.json]: o.O uh oh ... I didn't keep up for some it seems, didn't know it was already in stage 2.

**Comment by ta2edchimp** [comment]
Yes, technically there shouldn't be anything breaking; I guess I only wanted to "feel" safe after _"exit with error-gate"_ :stuck_out_tongue_winking_eye: 

Atm I have the following steps on my todo list for the upcoming days:

- [ ] Have a look into the code coverage issue.
- [ ] Re-/Implement the "exit with error" option.
- [ ] Update the ReadMe accordingly.

Please drop me a line here, if you'd prefer to put the (re-)implementation of "exit with error" into a separate PR (on the other hand, there would be no need for these changes, when there's nothing to configure).

**Comment by kentcdodds** [comment]
> Please drop me a line here, if you'd prefer to put the (re-)implementation of "exit with error" into a separate PR (on the other hand, there would be no need for these changes, when there's nothing to configure).

I'm fine doing this in the same PR as long as it's backward compatible :+1:

File: lib/hook.template.raw [hunk 4]
```diff
@@ -3,12 +3,21 @@
...
 if (checkForGHooks(ghooks)) {
   require(ghooks)(__dirname, __filename)
 }
 
+function getRenderedConfig() {
+  // ghooks' actual config gets rendered here
+  // it gets substituted on ghooks' setup
+  // {{generated_config_start}}
+  return global.__testonly_ghooksConfig__ || {}
+  // {{generated_config_end}}
```
Comment (ta2edchimp) [inline lib/hook.template.raw]: `global.__testonly_ghooksConfig__` is atm required to thoroughly test the optional behavior to exit with an error.

The whole block from `{{generated_config_start}}` to `{{generated_config_end}}` (including their comments) gets replaced when rendering the current config into the `hook.template`.

**As a side note:**
The `hooks` also get rendered into the template. I was unsure whether to ditch them, but maybe we can make use of them at a later point.

**Comment by gtramontina** [comment]
Hey everyone, really sorry for the silence… This looks great!

One other option, that just occurred me, could be simply having another `config` entry. Just like we have `config.ghooks`, I'm thinking we could also support `config.ghooks-settings` (or `config.ghooks-options`, or…) 💭 ?

> When going route 1, I'd suggest making the install script "re-runnable", by adding it as bin.ghooks to the package.json.

If I correctly got what you're saying, I believe we can already re-run it with `npm run ghooks install`.

**Comment by kentcdodds** [comment]
I'd be fine having a separate `config` property called `ghooksSettings` or whatever. I forget why we need settings for ghooks now though. It's been a while :-)

**Comment by gtramontina** [comment]
I thought an example would be useful… 😛 

```json
{
  "config": {
    "ghooks": {
      "pre-commit": "",
      …
    },
    "ghooks-settings": {
      "exitWithError": true
    }
  }
}
```
<sub>gonna sleep now, it's freaking 3am here o_O</sub>

**Comment by ta2edchimp** [comment]
>I forget why we need settings for ghooks now though. It's been a while :-)

It was mostly because of [»`exit with error`-gate«](https://github.com/gtramontina/ghooks/issues/70) and #28 (Specify project root manually).

Some place was concluded that it would be nice to eventually have

```JSON
"config": {
    "ghooks": {
        "randomConfig": ["randomValue"],
        "hooks": {
          "pre-commit": "echo \"some pre-commit hook\"
        }
    }
```

and to allow `config.ghooks-randomConfig` (whilst marked as deprecated) too — for reasons of backwards compatibility with `config.ghooks[ hookname ]`.

---

I should have some more time in the upcoming days, I'll at least update this PR / resolve the conflicts, so we'd have a clean status we'd be talking about ...

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 931c2ad0ea9cd82e621433226a831ea2a9a9df3b
  * Your commit message body contains a line that is longer than 72 characters

 * Commit: 2aefa30ea2828aa6e0bcb33879e93bc1ca8219df
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/gtramontina/ghooks

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 1547f50246fb0eb88962cad0ba44ac34be524941
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/gtramontina/ghooks

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 14e5ae7336e580a0f0be0c2f4b74557603b3cf51
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/gtramontina/ghooks

---
This message was auto-generated by https://gitcop.com

**Comment by ta2edchimp** [comment]
Jesus Christ, that GitCop tried hard to get me arrested for being too verbose . . . 😞

**Comment by gtramontina** [comment]
😆 I guess I should tune it down a little…

**Comment by gtramontina** [comment]
What do you think about warning the developer if they are using the deprecated (hopefully soon-to-be removed) way?

**Comment by ta2edchimp** [comment]
Absolutely! I'm just unsure as where to output the warning (as in, on every invokation of a hook, only on the hooks' installation/setup, ...).

**Comment by kentcdodds** [comment]
Re: GitCop, it seemed like a good idea at the time, but.... I don't use it anymore. Just too heavy handed and not flexible.

**Comment by gtramontina** [comment]
Closing this one as there will be no future development in `ghooks`. Thank you for your efforts, @ta2edchimp !

**Pull request closed without merging**